File: typescript/src/verify-export.ts
/**
 * Offline verifier for a signed AtlaSent audit-export bundle.
 *
 * Fed an envelope returned from {@link AtlaSentClient.exportAudit}
 * (or loaded from disk), this verifies:
 *
 *  1. For each row in the execution + admin chains,
 *     `sha256(canonical_payload) === entry_hash`.
 *  2. Each row's `canonical_payload` ends with the previous row's
 *     `entry_hash` (or `"GENESIS"` for row 0).
 *  3. The Ed25519 signature over
 *     `canonicalize(envelope - signature)` verifies against the
 *     embedded `public_key_pem`.
 *  4. When `trustedPublicKeyPem` is supplied, that the embedded key
 *     matches it byte-for-byte — so the verifier trusts a key **you**
 *     provisioned, not one shipped inside the envelope.
 *
 * Zero runtime dependencies — uses `node:crypto` only. Node 20+.
 */

import {
  createPublicKey,
  verify as nodeVerify,
  type KeyObject,
} from "node:crypto";
import { readFile } from "node:fs/promises";

import { canonicalize, sha256Hex } from "./canonical.js";

export interface VerifyBundleResult {
  /** Every chain row hashes to its `entry_hash` and links to the previous. */
  chainOk: boolean;
  /** The Ed25519 signature verifies against the embedded public key. */
  signatureOk: boolean;
  /**
   * `true` when the embedded public key matches `trustedPublicKeyPem`.
   * `null` when no trust anchor was supplied (so not checked).
   */
  trustedKeyOk: boolean | null;
  /** Human-readable diagnostics — empty on a clean verify. */
  errors: string[];
  /**
   * `true` iff `chainOk && signatureOk && (trustedKeyOk !== false)`.
   */
  ok: boolean;
}

export interface VerifyBundleOptions {
  /**
   * The Ed25519 public key you trust, provisioned out of band. When
   * supplied, the verifier asserts the envelope's embedded key matches
   * byte-for-byte.
   */
  trustedPublicKeyPem?: string;
}

/**
 * Verify a signed audit-export envelope without hitting the API.
 *
 * Accepts either the parsed envelope object (from
 * `bundle.raw`, `JSON.parse(...)`, etc.) or a filesystem path to a
 * JSON file.
 */
export async function verifyBundle(
  input: Record<string, unknown> | string,
  options: VerifyBundleOptions = {},
): Promise<VerifyBundleResult> {
  let envelope: Record<string, unknown>;
  if (typeof input === "string") {
    const text = await readFile(input, "utf8");
    envelope = JSON.parse(text) as Record<string, unknown>;
  } else if (input !== null && typeof input === "object") {
    envelope = input;
  } else {
    throw new TypeError(
      "verifyBundle: input must be an envelope object or a path string",
    );
  }
  return verifyEnvelope(envelope, options.trustedPublicKeyPem);
}

function verifyEnvelope(
  envelope: Record<string, unknown>,
  trustedPublicKeyPem?: string,
): VerifyBundleResult {
  const errors: string[] = [];

  const executionRows = asRowArray(envelope.evaluations);
  const adminRows = asRowArray(envelope.admin_log);
  const chainOk =
    verifyChain(executionRows, envelope.execution_head, "execution", errors) &&
    verifyChain(adminRows, envelope.admin_head, "admin", errors);

  const signature = envelope.signature;
  const pem = envelope.public_key_pem;
  if (
    typeof signature !== "string" ||
    signature.length === 0 ||
    typeof pem !== "string" ||
    pem.length === 0
  ) {
    errors.push("missing signature or public_key_pem");
    return finalize(chainOk, false, null, errors);
  }

  let embeddedKey: KeyObject;
  try {
    embeddedKey = createPublicKey(pem);
  } catch (err) {
    errors.push(
      `could not parse embedded public key: ${(err as Error).message}`,
    );
    return finalize(chainOk, false, null, errors);
  }

  const envMinusSig: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(envelope)) {
    if (k !== "signature") envMinusSig[k] = v;
  }
  const canonicalBytes = Buffer.from(canonicalize(envMinusSig), "utf8");

  let sigBytes: Buffer;
  try {
    sigBytes = Buffer.from(signature, "base64");
  } catch (err) {
    errors.push(`signature is not valid base64: ${(err as Error).message}`);
    return finalize(chainOk, false, null, errors);
  }

  let signatureOk = false;
  try {
    signatureOk = nodeVerify(null, canonicalBytes, embeddedKey, sigBytes);
  } catch (err) {
    errors.push(
      `signature verification threw: ${(err as Error).message}`,
    );
    signatureOk = false;
  }
  if (!signatureOk) {
    errors.push("signature does not verify against embedded public key");
  }

  let trustedKeyOk: boolean | null = null;
  if (trustedPublicKeyPem !== undefined) {
    trustedKeyOk = pemNormalize(trustedPublicKeyPem) === pemNormalize(pem);
    if (!trustedKeyOk) {
      errors.push("embedded public key does not match the trusted anchor");
    }
  }

  return finalize(chainOk, signatureOk, trustedKeyOk, errors);
}

function finalize(
  chainOk: boolean,
  signatureOk: boolean,
  trustedKeyOk: boolean | null,
  errors: string[],
): VerifyBundleResult {
  return {
    chainOk,
    signatureOk,
    trustedKeyOk,
    errors,
    ok: chainOk && signatureOk && trustedKeyOk !== false,
  };
}

function asRowArray(value: unknown): Array<Record<string, unknown>> {
  if (!Array.isArray(value)) return [];
  return value as Array<Record<string, unknown>>;
}

function verifyChain(
  rows: Array<Record<string, unknown>>,
  claimedHead: unknown,
  label: string,
  errors: string[],
): boolean {
  let prev: string | null = null;
  for (let i = 0; i < rows.length; i++) {
    const row = rows[i]!;
    const payload = row.canonical_payload;
    const stored = row.entry_hash;
    if (typeof payload !== "string" || typeof stored !== "string") {
      errors.push(
        `${label} row ${i}: missing canonical_payload or entry_hash`,
      );
      return false;
    }
    if (sha256Hex(payload) !== stored) {
      errors.push(
        `${label} row ${i} (${String(
          row.id,
        )}): sha256(canonical_payload) !== entry_hash`,
      );
      return false;
    }
    const segments = payload.split("|");
    const payloadPrev = segments[segments.length - 1] ?? "";
    const expectedPrev = prev ?? "GENESIS";
    if (payloadPrev !== expectedPrev) {
      errors.push(
        `${label} row ${i}: payload prev '${payloadPrev}' !== '${expectedPrev}'`,
      );
      return false;
    }
    prev = stored;
  }

  if (
    rows.length > 0 &&
    claimedHead &&
    typeof claimedHead === "object" &&
    (claimedHead as Record<string, unknown>).entry_hash !== prev
  ) {
    errors.push(
      `${label}: claimed head ${String(
        (claimedHead as Record<string, unknown>).entry_hash,
      )} does not match tail ${prev}`,
    );
    return false;
  }
  return true;
}

function pemNormalize(pem: string): string {
  return pem
    .replace(/-----BEGIN PUBLIC KEY-----/g, "")
    .replace(/-----END PUBLIC KEY-----/g, "")
    .replace(/\r/g, "")
    .replace(/\n/g, "")
    .replace(/ /g, "")
    .trim();
}

File: python/atlasent/verify_export.py
"""Offline verifier for a signed AtlaSent audit-export bundle.

Fed an envelope returned from :meth:`AtlaSentClient.export_audit`
(or loaded from disk), this verifies:

1. For each row in the execution + admin chains:
   ``sha256(canonical_payload) == entry_hash``.
2. Each row's ``canonical_payload`` ends with the previous row's
   ``entry_hash`` (or ``'GENESIS'`` for row 0).
3. The Ed25519 signature over
   ``canonicalize(envelope - signature)`` verifies against the
   embedded ``public_key_pem``.
4. When ``trusted_public_key_pem`` is supplied, that the embedded key
   matches it byte-for-byte — so the verifier trusts a key **you**
   provisioned, not one shipped inside the envelope.

Requires the optional ``cryptography`` dependency. Install with::

    pip install 'atlasent[verify]'
"""

from __future__ import annotations

import base64
import json
from collections.abc import Mapping
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .canonical import canonicalize, sha256_hex


@dataclass
class ExportVerifyResult:
    """Outcome of :func:`verify_bundle` / :func:`verify_audit_export`.

    ``ok`` is ``True`` iff the chain and signature both verify, and
    (when a trust anchor was supplied) the embedded key matches it.
    """

    chain_ok: bool
    signature_ok: bool
    trusted_key_ok: bool | None = None
    errors: list[str] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return (
            self.chain_ok
            and self.signature_ok
            and (self.trusted_key_ok is None or self.trusted_key_ok)
        )


def verify_bundle(
    envelope: Mapping[str, Any] | str | Path,
    *,
    trusted_public_key_pem: str | None = None,
) -> ExportVerifyResult:
    """Verify a signed audit-export envelope without hitting the API.

    Args:
        envelope: Either the parsed envelope dict (e.g.
            ``AuditExportBundle.model_dump()`` or ``bundle.raw``), or
            a path to a JSON file on disk.
        trusted_public_key_pem: The Ed25519 public key you trust,
            provisioned out of band. When supplied, the verifier
            asserts the envelope's embedded key matches byte-for-byte.

    Returns:
        :class:`ExportVerifyResult`. Check ``.ok`` for the final
        answer; ``.errors`` holds human-readable diagnostics.
    """
    if isinstance(envelope, (str, Path)):
        with open(envelope, encoding="utf-8") as f:
            envelope = json.load(f)
    if not isinstance(envelope, Mapping):
        raise TypeError(
            "verify_bundle: envelope must be a mapping, str, or Path "
            f"(got {type(envelope).__name__})"
        )
    return _verify(envelope, trusted_public_key_pem)


# Back-compat alias matching the atlasent-api reference SDK.
verify_audit_export = verify_bundle


def _verify(
    envelope: Mapping[str, Any],
    trusted_public_key_pem: str | None,
) -> ExportVerifyResult:
    result = ExportVerifyResult(chain_ok=False, signature_ok=False)

    chain_ok = _verify_chain(
        envelope.get("evaluations") or [],
        envelope.get("execution_head"),
        "execution",
        result.errors,
    ) and _verify_chain(
        envelope.get("admin_log") or [],
        envelope.get("admin_head"),
        "admin",
        result.errors,
    )
    result.chain_ok = chain_ok

    signature = envelope.get("signature")
    pem = envelope.get("public_key_pem")
    if not signature or not pem:
        result.errors.append("missing signature or public_key_pem")
        return result

    try:
        public_key = _load_ed25519_public_key(str(pem))
    except Exception as err:  # noqa: BLE001 — propagate as a verify failure
        result.errors.append(f"could not parse embedded public key: {err}")
        return result

    env_minus_sig = {k: v for k, v in envelope.items() if k != "signature"}
    canonical_bytes = canonicalize(env_minus_sig).encode("utf-8")
    try:
        sig_bytes = base64.b64decode(str(signature), validate=False)
    except Exception as err:  # noqa: BLE001
        result.errors.append(f"signature is not valid base64: {err}")
        return result

    try:
        public_key.verify(sig_bytes, canonical_bytes)
        result.signature_ok = True
    except Exception as err:  # noqa: BLE001 — InvalidSignature or anything else
        result.errors.append(
            f"signature does not verify against embedded public key: {err}"
        )

    if trusted_public_key_pem is not None:
        trusted_ok = _pem_normalize(trusted_public_key_pem) == _pem_normalize(str(pem))
        result.trusted_key_ok = trusted_ok
        if not trusted_ok:
            result.errors.append(
                "embedded public key does not match the trusted anchor"
            )

    return result


def _verify_chain(
    rows: list[Any],
    claimed_head: Any,
    label: str,
    errors: list[str],
) -> bool:
    prev: str | None = None
    for i, row in enumerate(rows):
        if not isinstance(row, Mapping):
            errors.append(f"{label} row {i}: expected an object")
            return False
        payload = row.get("canonical_payload")
        stored = row.get("entry_hash")
        if not isinstance(payload, str) or not isinstance(stored, str):
            errors.append(
                f"{label} row {i}: missing canonical_payload or entry_hash"
            )
            return False

        if sha256_hex(payload) != stored:
            errors.append(
                f"{label} row {i} ({row.get('id')}): "
                f"sha256(canonical_payload) != entry_hash"
            )
            return False

        segments = payload.split("|")
        payload_prev = segments[-1] if segments else ""
        expected_prev = prev if prev is not None else "GENESIS"
        if payload_prev != expected_prev:
            errors.append(
                f"{label} row {i}: payload prev '{payload_prev}' "
                f"!= '{expected_prev}'"
            )
            return False
        prev = stored

    if (
        claimed_head
        and rows
        and isinstance(claimed_head, Mapping)
        and claimed_head.get("entry_hash") != prev
    ):
        errors.append(
            f"{label}: claimed head {claimed_head.get('entry_hash')} "
            f"does not match tail {prev}"
        )
        return False
    return True


def _load_ed25519_public_key(pem: str) -> Any:
    try:
        from cryptography.hazmat.primitives.serialization import load_pem_public_key
    except ImportError as err:  # pragma: no cover — guarded by extras
        raise RuntimeError(
            "verify_bundle requires the 'cryptography' package. "
            "Install with: pip install 'atlasent[verify]'"
        ) from err
    return load_pem_public_key(pem.encode("utf-8"))


def _pem_normalize(pem: str) -> str:
    """Strip PEM armour + whitespace for byte-equality comparison."""
    return (
        pem.replace("-----BEGIN PUBLIC KEY-----", "")
        .replace("-----END PUBLIC KEY-----", "")
        .replace("\r", "")
        .replace("\n", "")
        .replace(" ", "")
        .strip()
    )
